File: project_euler_problems/solved_problems/cpp_solutions/problem_10/10.cpp
#include <iostream>
using namespace std;

/* Project Euler #10:
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 *
 * Find the sum of all the primes below two million. */

void markIndex(bool sieve[], int num, const int size) {
    for (int i = (num * 2); i < size; i += num) {
        sieve[i] = false;
    }
}

int main()
{
    const int SIZE = 2000000;
    bool sieve[SIZE];
    for (int i = 0; i < SIZE; i++) {
        sieve[i] = true;
    }
    for (int i = 2; i < SIZE; i++) {
        if(sieve[i]) {
            markIndex(sieve, i, SIZE);
        }
    }
    long sum = 0;
    for (int i = 2; i < SIZE; i++) {
        if (sieve[i]) {
            sum += i;
        }
    }
    // Answer: 142913828922
    cout << "Sum of all primes < 2 million: "
         << sum << endl;
    return 0;
}

File: project_euler_problems/solved_problems/python_solutions/10.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

"""
Project Euler Problem #10:

The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
Find the sum of all the primes below two million.
"""

# Algorithm: Sieve of Eratosthenes:
# Deletes the multiples of all primes upto a given number.
import time

start = time.time()
Sieve = [True] * 2000000

def mark(Sieve, X):
    for i in xrange(X * 2, len(Sieve), X): # Loops for all multiples.
        Sieve[i] = False # Deletes the possibility.

for x in xrange(2, len(Sieve)):
    if Sieve[x]:
        mark(Sieve, x)

print "Sum of all primes < 2m:", sum([i for i in xrange(2, len(Sieve)) if Sieve[i]]) #Answer: 142913828922
print "Time to finish: ", str(time.time() - start), "seconds."
